Mixpanel/Mixpanel.m

```diff
@@ -694,51 +706,48 @@ - (double)eventElapsedTime:(NSString *)event {
+- (void)resetWithCompletion:(void (^)(void))handler
 {
     [self flush];
-    dispatch_async(self.serialQueue, ^{
+    dispatch_async(self.backgroundQueue, ^{
         // wait for all current network requests to finish before resetting
         dispatch_sync(self.networkQueue, ^{ return; });
-        @synchronized (self) {
-            self.distinctId = [self defaultDistinctId];
-            self.superProperties = [NSMutableDictionary dictionary];
-            self.people.distinctId = nil;
-            self.alias = nil;
-            self.people.unidentifiedQueue = [NSMutableArray array];
-            self.eventsQueue = [NSMutableArray array];
-            self.peopleQueue = [NSMutableArray array];
-            self.timedEvents = [NSMutableDictionary dictionary];
-            self.shownNotifications = [NSMutableSet set];
-            self.decideResponseCached = NO;
-            self.variants = [NSSet set];
-            self.eventBindings = [NSSet set];
+        dispatch_sync(self.backgroundQueue, ^{
+            @synchronized (self) {
+                self.distinctId = [self defaultDistinctId];
+                self.superProperties = [NSMutableDictionary dictionary];
+                self.people.distinctId = nil;
+                self.alias = nil;
+                self.people.unidentifiedQueue = [NSMutableArray array];
+                self.eventsQueue = [NSMutableArray array];
+                self.peopleQueue = [NSMutableArray array];
+                self.timedEvents = [NSMutableDictionary dictionary];
+                self.shownNotifications = [NSMutableSet set];
+                self.decideResponseCached = NO;
+                self.variants = [NSSet set];
+                self.eventBindings = [NSSet set];
 #if !MIXPANEL_NO_CONNECT_INTEGRATION_SUPPORT
-            [self.connectIntegrations reset];
+                [self.connectIntegrations reset];
 #endif
 #if !MIXPANEL_NO_NOTIFICATION_AB_TEST_SUPPORT
-            if (![Mixpanel isAppExtension]) {
-                [[MPTweakStore sharedInstance] reset];
-            }
+                if (![Mixpanel isAppExtension]) {
+                    [[MPTweakStore sharedInstance] reset];
+                }
 #endif
-        }
-        [self archive];
+            }
+            [self archive];
+            dispatch_async(dispatch_get_main_queue(), handler);
+        });
     });
 }
 
 - (void)dispatchOnNetworkQueue:(void (^)(void))dispatchBlock {
-    // so this looks stupid but to make [Mixpanel track]; [Mixpanel flush]; continue to have the track
-    // guaranteed to be part of the flush we need to make networkQueue stuff be dispatched on serialQueue
-    // first. still will allow serialQueue stuff to happen at the same time as networkQueue stuff just
-    // don't want to change track -> flush behavior that people may be relying on
-    dispatch_async(self.serialQueue, ^{
-        dispatch_async(self.networkQueue, dispatchBlock);
-    });
+    dispatch_async(self.networkQueue, dispatchBlock);
```

Comment:
I could be missing something but I think using `dispatch_async` allows this to be simplified.